File: src/app/core/services/weather.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { catchError, map, Observable, switchMap, throwError } from 'rxjs';

export type WeatherType =
  | 'sunny'
  | 'partly-cloudy'
  | 'cloudy'
  | 'rainy'
  | 'heavy-rain';

export interface ForecastDay {
  day: string;
  date: string;
  minTemperature: number;
  maxTemperature: number;
  rainChance: number;
  condition: string;
  weatherType: WeatherType;
  icon: string;
}

export interface WeatherData {
  cityName: string;
  country: string;
  countryCode: string;
  temperature: number;
  feelsLike: number;
  humidity: number;
  windSpeed: number;
  condition: string;
  weatherType: WeatherType;
  icon: string;
  updatedAt: string;
  forecast: ForecastDay[];
}

interface GeocodingResponse {
  results?: GeocodingResult[];
}

interface GeocodingResult {
  name: string;
  latitude: number;
  longitude: number;
  country: string;
  country_code: string;
  timezone: string;
  admin1?: string;
}

interface OpenMeteoResponse {
  current: {
    time: string;
    temperature_2m: number;
    relative_humidity_2m: number;
    apparent_temperature: number;
    weather_code: number;
    wind_speed_10m: number;
  };
  daily: {
    time: string[];
    weather_code: number[];
    temperature_2m_max: number[];
    temperature_2m_min: number[];
    precipitation_probability_max: number[];
  };
}

@Injectable({
  providedIn: 'root'
})
export class WeatherService {
  private readonly geocodingUrl = 'https://geocoding-api.open-meteo.com/v1/search';
  private readonly forecastUrl = 'https://api.open-meteo.com/v1/forecast';

  constructor(private http: HttpClient) {}

  getWeatherByCity(city: string): Observable<WeatherData> {
    const geocodingParams = new HttpParams()
      .set('name', city)
      .set('count', 1)
      .set('language', 'pt')
      .set('format', 'json');

    return this.http
      .get<GeocodingResponse>(this.geocodingUrl, { params: geocodingParams })
      .pipe(
        map((response) => {
          const location = response.results?.[0];

          if (!location) {
            throw new Error('Cidade não encontrada.');
          }

          return location;
        }),
        switchMap((location) => this.getForecastByLocation(location)),
        catchError((error) => {
          const message =
            error?.message || 'Não foi possível buscar o clima agora.';

          return throwError(() => new Error(message));
        })
      );
  }

  private getForecastByLocation(
    location: GeocodingResult
  ): Observable<WeatherData> {
    const forecastParams = new HttpParams()
      .set('latitude', location.latitude)
      .set('longitude', location.longitude)
      .set(
        'current',
        'temperature_2m,relative_humidity_2m,apparent_temperature,weather_code,wind_speed_10m'
      )
      .set(
        'daily',
        'weather_code,temperature_2m_max,temperature_2m_min,precipitation_probability_max'
      )
      .set('timezone', 'auto')
      .set('forecast_days', 7)
      .set('wind_speed_unit', 'kmh');

    return this.http
      .get<OpenMeteoResponse>(this.forecastUrl, { params: forecastParams })
      .pipe(map((response) => this.mapToWeatherData(location, response)));
  }

  private mapToWeatherData(
    location: GeocodingResult,
    response: OpenMeteoResponse
  ): WeatherData {
    const currentCode = response.current.weather_code;
    const currentWeatherType = this.getWeatherType(currentCode);

    return {
      cityName: location.name,
      country: location.country,
      countryCode: location.country_code,
      temperature: Math.round(response.current.temperature_2m),
      feelsLike: Math.round(response.current.apparent_temperature),
      humidity: response.current.relative_humidity_2m,
      windSpeed: Math.round(response.current.wind_speed_10m),
      condition: this.getCondition(currentCode),
      weatherType: currentWeatherType,
      icon: this.getIcon(currentWeatherType),
      updatedAt: response.current.time,
      forecast: this.mapForecast(response)
    };
  }

  private mapForecast(response: OpenMeteoResponse): ForecastDay[] {
    return response.daily.time.slice(0, 7).map((date, index) => {
      const weatherCode = response.daily.weather_code[index];
      const weatherType = this.getWeatherType(weatherCode);

      return {
        date,
        day: this.formatDay(date),
        minTemperature: Math.round(response.daily.temperature_2m_min[index]),
        maxTemperature: Math.round(response.daily.temperature_2m_max[index]),
        rainChance: response.daily.precipitation_probability_max[index] ?? 0,
        condition: this.getCondition(weatherCode),
        weatherType,
        icon: this.getIcon(weatherType)
      };
    });
  }

  private formatDay(date: string): string {
    return new Intl.DateTimeFormat('en-US', {
      weekday: 'short'
    }).format(new Date(`${date}T00:00:00`));
  }

  private getWeatherType(code: number): WeatherType {
    if (code === 0 || code === 1) {
      return 'sunny';
    }

    if (code === 2) {
      return 'partly-cloudy';
    }

    if (code === 3 || code === 45 || code === 48) {
      return 'cloudy';
    }

    if (
      code === 51 ||
      code === 53 ||
      code === 55 ||
      code === 56 ||
      code === 57 ||
      code === 61 ||
      code === 63 ||
      code === 65 ||
      code === 66 ||
      code === 67 ||
      code === 80 ||
      code === 81
    ) {
      return 'rainy';
    }

    if (code === 82 || code === 95 || code === 96 || code === 99) {
      return 'heavy-rain';
    }

    return 'cloudy';
  }

  private getCondition(code: number): string {
    const conditions: Record<number, string> = {
      0: 'Sunny',
      1: 'Mostly sunny',
      2: 'Partly cloudy',
      3: 'Cloudy',
      45: 'Foggy',
      48: 'Foggy',
      51: 'Light drizzle',
      53: 'Drizzle',
      55: 'Heavy drizzle',
      56: 'Freezing drizzle',
      57: 'Freezing drizzle',
      61: 'Light rain',
      63: 'Rainy',
      65: 'Heavy rain',
      66: 'Freezing rain',
      67: 'Freezing rain',
      71: 'Light snow',
      73: 'Snowy',
      75: 'Heavy snow',
      77: 'Snow grains',
      80: 'Light showers',
      81: 'Rain showers',
      82: 'Heavy rain',
      85: 'Snow showers',
      86: 'Heavy snow showers',
      95: 'Thunderstorm',
      96: 'Thunderstorm with hail',
      99: 'Severe thunderstorm'
    };

    return conditions[code] ?? 'Cloudy';
  }

  private getIcon(weatherType: WeatherType): string {
    return `assets/weather-icons/${weatherType}.svg`;
  }
}

File: onlyweather-api/src/main/java/com/onlyweather/api/service/WeatherService.java
package com.onlyweather.api.service;

import com.onlyweather.api.client.GeocodingClient;
import com.onlyweather.api.client.OpenMeteoClient;
import com.onlyweather.api.dto.CityLocationResponse;
import com.onlyweather.api.dto.ForecastDayResponse;
import com.onlyweather.api.dto.OpenMeteoApiResponse;
import com.onlyweather.api.dto.WeatherResponse;
import org.springframework.stereotype.Service;
import java.time.LocalDate;
import java.time.format.TextStyle;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

@Service
public class WeatherService {

    private final GeocodingClient geocodingClient;
    private final OpenMeteoClient openMeteoClient;

    public WeatherService(
            GeocodingClient geocodingClient,
            OpenMeteoClient openMeteoClient
    ) {
        this.geocodingClient = geocodingClient;
        this.openMeteoClient = openMeteoClient;
    }

    public WeatherResponse getWeatherByCity(String city) {
        CityLocationResponse location = geocodingClient.findCity(city);
        OpenMeteoApiResponse apiResponse = openMeteoClient.getForecast(location);

        if (apiResponse == null || apiResponse.current() == null || apiResponse.daily() == null) {
            throw new IllegalStateException("Resposta inválida da API de clima.");
        }

        int currentCode = apiResponse.current().weatherCode();
        boolean isDay = apiResponse.current().isDay() == 1;

        String weatherType = getCurrentWeatherType(currentCode, isDay);

        return new WeatherResponse(
                location.name(),
                location.country(),
                location.countryCode(),
                location.latitude(),
                location.longitude(),
                round(apiResponse.current().temperature()),
                round(apiResponse.current().apparentTemperature()),
                apiResponse.current().humidity(),
                round(apiResponse.current().windSpeed()),
                getCondition(currentCode, isDay),
                weatherType,
                getIcon(weatherType),
                apiResponse.current().time(),
                mapForecast(apiResponse)
        );
    }

    private List<ForecastDayResponse> mapForecast(OpenMeteoApiResponse apiResponse) {
        List<ForecastDayResponse> forecast = new ArrayList<>();
        List<String> dates = apiResponse.daily().time();

        for (int index = 0; index < dates.size(); index++) {
            int weatherCode = apiResponse.daily().weatherCode().get(index);
            String weatherType = getDayWeatherType(weatherCode);
            String date = dates.get(index);

            forecast.add(
                    new ForecastDayResponse(
                            formatDay(date),
                            date,
                            round(apiResponse.daily().minTemperature().get(index)),
                            round(apiResponse.daily().maxTemperature().get(index)),
                            getSafeRainChance(apiResponse, index),
                            getCondition(weatherCode, true), weatherType, 
                            getIcon(weatherType)
                    )
            );
        }

        return forecast;
    }

    private int getSafeRainChance(OpenMeteoApiResponse apiResponse, int index) {
        if (apiResponse.daily().rainChance() == null) {
            return 0;
        }

        Integer value = apiResponse.daily().rainChance().get(index);

        return value == null ? 0 : value;
    }

    private String formatDay(String date) {
        LocalDate localDate = LocalDate.parse(date);

        return localDate.getDayOfWeek().getDisplayName(TextStyle.SHORT, Locale.ENGLISH);
    }

    private String getCurrentWeatherType(int code, boolean isDay) {
        if (isDay) {
            return getDayWeatherType(code);
        }

        if (isStorm(code)) {
            return "stormy-night";
        }

        if (isRain(code)) {
            return "rainy-night";
        }

        return "night";
    }

    private String getDayWeatherType(int code) {
        if (code == 0 || code == 1) {
            return "sunny";
        }

        if (code == 2) {
            return "partly-cloudy";
        }

        if (isCloudy(code)) {
            return "cloudy";
        }

        if (isRain(code)) {
            return "rainy";
        }

        if (isStorm(code)) {
            return "heavy-rain";
        }

        return "cloudy";
    }

    private boolean isCloudy(int code) {
        return code == 3 || code == 45 || code == 48;
    }

    private boolean isRain(int code) {
        return code == 51 ||
                code == 53 ||
                code == 55 ||
                code == 56 ||
                code == 57 ||
                code == 61 ||
                code == 63 ||
                code == 65 ||
                code == 66 ||
                code == 67 ||
                code == 80 ||
                code == 81;
    }

    private boolean isStorm(int code) {
        return code == 82 ||
                code == 95 ||
                code == 96 ||
                code == 99;
    }

    private String getCondition(int code, boolean isDay) {
        if (!isDay) {
            if (isStorm(code)) {
                return "Stormy night";
            }

            if (isRain(code)) {
                return "Rainy night";
            }

            return "Night";
        }

        return switch (code) {
            case 0 -> "Sunny";
            case 1 -> "Mostly sunny";
            case 2 -> "Partly cloudy";
            case 3 -> "Cloudy";
            case 45, 48 -> "Foggy";
            case 51 -> "Light drizzle";
            case 53 -> "Drizzle";
            case 55 -> "Heavy drizzle";
            case 56, 57 -> "Freezing drizzle";
            case 61 -> "Light rain";
            case 63 -> "Rainy";
            case 65 -> "Heavy rain";
            case 66, 67 -> "Freezing rain";
            case 71 -> "Light snow";
            case 73 -> "Snowy";
            case 75 -> "Heavy snow";
            case 77 -> "Snow grains";
            case 80 -> "Light showers";
            case 81 -> "Rain showers";
            case 82 -> "Heavy rain";
            case 85 -> "Snow showers";
            case 86 -> "Heavy snow showers";
            case 95 -> "Thunderstorm";
            case 96 -> "Thunderstorm with hail";
            case 99 -> "Severe thunderstorm";
            default -> "Cloudy";
        };
    }

    private String getIcon(String weatherType) {
        return "assets/weather-icons/" + weatherType + ".svg";
    }

    private int round(double value) {
        return (int) Math.round(value);
    }
}